dimod/serialization/utils.py

```diff
@@ -13,9 +13,88 @@
 #    limitations under the License.
 #
 # =============================================================================
+
+try:
+    import collections.abc as abc
+except ImportError:
+    import collections as abc
+
 import numpy as np
 
 
+def serialize_ndarray(arr, use_bytes=False, bytes_type=bytes):
+    """Serializes a NumPy array.
+
+    Args:
+        arr (array-like):
+            An array.
+
+        use_bytes (bool, optional, default=False):
+            If True, a compact representation of the biases as bytes is used.
+
+        bytes_type (class, optional, default=bytes):
+            If `use_bytes` is True, this class is used to wrap the bytes
+            objects in the serialization. Useful for Python 2 using BSON
+            encoding, which does not accept the raw `bytes` type;
+            `bson.Binary` can be used instead.
+
+    Returns:
+        dict: A serializable object.
+
+    """
+    arr = np.asarray(arr)  # support array-like
+    if use_bytes:
+        data = bytes_type(arr.tobytes(order='C'))
+    else:
+        data = arr.tolist()
+    return dict(type='array',
+                data=data,
+                data_type=arr.dtype.name,
+                shape=arr.shape,
+                use_bytes=bool(use_bytes))
+
+
+def deserialize_ndarray(obj):
+    """Inverse of serialize_ndarray.
+
+    Args:
+        obj (dict):
+            As constructed by :func:`.serialize_ndarray`.
+
+    Returns:
+        :obj:`numpy.ndarray`
+
+    """
+    if obj['use_bytes']:
+        arr = np.frombuffer(obj['data'], dtype=obj['data_type'])
+    else:
+        arr = np.asarray(obj['data'], dtype=obj['data_type'])
+    arr = arr.reshape(obj['shape'])  # makes a view generally, but that's fine
+    return arr
+
+
+def dfs_serialize_ndarray(obj, use_bytes=False, bytes_type=bytes):
+    """Looks through the object, serializing numpy arrays."""
+    if isinstance(obj, np.ndarray):
+        return serialize_ndarray(obj, use_bytes=use_bytes, bytes_type=bytes_type)
+    elif isinstance(obj, abc.Mapping):
+        return {key: dfs_serialize_ndarray(val) for key, val in obj.items()}
```

Comment:
Keys can be NumPy types too.

Reply:
Not if it's JSON serializable. Unless numpy chars?

Reply:
Also numpy numbers.

Reply:
JSON needs string keys for objects. I don't think we should handle that case here.

Reply:
Also seems to work for the numeric types out of the box
```
In [4]: json.dumps({np.float(0): 1})                                                                                                   
Out[4]: '{"0.0": 1}'
```

Reply:
Yes, I know JSON requires string keys. But as you see, built-in json package serializes python numbers as strings. It fails with numpy though:
```
>>> json.dumps({np.int64(0): 1})
...
TypeError: keys must be a string
```

Reply:
I'm not saying I like casting arrays to strings, but this solution handles only ndarrays in values. I feel if we already are providing this service to users (serialization of numpy types), we should be little broader.

Reply:
It's a simple enough change so will do. However, I don't think it is in scope for dimod to handle serialization of numpy objects in general. So far we are only providing it as an addition to the serialization of dimod objects. If we want to go beyond that this code will need to be a lot more sophisticated.

Reply:
I agree. Actually, if you remember, I argued even more than that -- BQM/SampleSet serialization (for SAPI encoding) does not belong in dimod at all.

Reply:
I think serialization of dimod objects makes sense in dimod. The numpy objects that happen to be stored less so.

Reply:
(But as long as we have it already, and we're now adding NumPy support too, I'm saying let's just cover simple edge cases to minimize gotcha moments.)